Users/randyh/hiptests doc (#3676)
ROCm/rocm-systems: projects/hip/docs/tutorial/hiptests.rst

.. meta::
  :description: The SAXPY tutorial on HIP
  :keywords: AMD, ROCm, HIP, SAXPY, tutorial

*******************************************************************************
Working with HIP Tests
*******************************************************************************

This tutorial introduces the HIP tests component available alongside the HIP API and runtime.
HIP tests are not installed as part of the standard ROCm installation, but you
can build ``rocm-systems/projects/hip-tests`` as described in :ref:`build-tests`.

The catch tests and samples provided in ``hip-tests`` are both a diagnostic tool and a learning toolkit.
You can use the tools and tests to validate your own system environment and HIP installation. You can

> also review the code and see how specific aspecis of the HIP API are used under diferent conditions. 

Reviewer: ```suggestion
You can use the tools and tests to validate your own system environment and HIP installation. You can
also review the code and see how specific aspects of the HIP API are used under different conditions. 
```

Running HIP tests
=================

> The HIP tests are catch tests provided alongside the HIP component, letting you test your
> system and the HIP runtime. In addition HIP tests provides some sample applications and
> utilities that can be individually built and run as needed. 

Reviewer: ```suggestion
The HIP tests are catch tests provided alongside the HIP component, letting you test your
system and the HIP runtime. In addition, HIP tests provide some sample applications and
utilities that can be individually built and run as needed. 
```

and sample suite for HIP. The ``rocm-systems/projects/hip-tests/catch`` folder contains the lightweight Catch2-based
C++ testing library unit test framework used by HIP tests. It provides the test harness for HIP API tests and lets
you see how HIP functions are used and validated in isolation.

The catch tests of ``hip-tests`` serves two complementary purposes: provides a comprehensive diagnostic suite for your
HIP installation, and executable examples of how to write catch tests for your own code. The tests show:

* How to structure a HIP unit test
* How to wrap HIP API calls with assertions
* How to test kernels deterministically
* How to test memory behavior, events, streams, and error handling
* How to isolate failures cleanly
* How to write reproducible GPU tests that don’t depend on timing or scheduling quirks

Follow the instructions in :ref:`build-tests`, and run the tests using the ``ctest`` command from the ``build`` folder. The command runs over four thousand tests and returns results in the ``build/Testing/Temporary/LastTest.log`` file. You can also build and run individual tests as explained in the build instructions. 

> 

Reviewer: ```suggestion
Follow the instructions in :ref:`build-tests`, and run the tests using the ``ctest`` command from the ``build`` folder. The command runs more than 4,000 tests and writes results in the ``build/Testing/Temporary/LastTest.log`` file. You can also build and run individual tests as explained in the build instructions. 

```

Building and running HIP samples
================================

> The ROCm/hip-tests/samples folder is organized into a set of small utilities that show how to use HIP APIs. Each sample is a utility that illustrates a concept, pattern, or technique. The folder is divided into three major groups:

Reviewer: ```suggestion
The ROCm/hip-tests/samples folder is organized into a set of small utilities that demonstrate how to use HIP APIs. Each sample is a utility that illustrates a concept, pattern, or technique. The folder is divided into three major groups:
```

  - ``samples/0_Intro/module_api_global``: Extends the module API example to show how to access global variables defined inside a GPU module, use global variables to configure the kernel, and copy data to and from those global variables. 
  - ``samples/0_Intro/square``: Implements a small kernel that squares each element of an array. It demonstrates minimal kernel structure, memory allocation, launch parameters, and verifying results on the host. 

* ``samples/1_Utils``: Provides reusable helper utilities and patterns.

  - ``samples/1_Utils/hipDispatchLatency``: This utility benchmarks the latency of dispatching a HIP kernel, focusing on the performance cost of launching kernels rather than the cost of running them.
  - ``samples/1_Utils/hipInfo``: Provides information about the HIP runtime and the available GPU devices. ``hipInfo`` is similar to ``rocminfo``, but returns only what the HIP runtime exposes via ``hipGetDeviceProperties()`` and related APIs. 

> * ``samples/2_Cookbook``: Demonstrates more advanced examples and build techniques. The concepts covered by the Cookbook include a variety of operations such as matrix operations, memory‑access patterns, warp/wavefront shuffle and data exchange, streams, concurrency, and multi‑GPU, HIP events and timing, occupancy, architecture, and device‑specific behavior. The following is a brief description of the available modules. 

Reviewer: Do we want to use both warp and wavefront?

>   - ``samples/2_Cookbook/0_MatrixTranspose``: Demonstrates implementing and optimizing a matrix transpose kernel. Helps to understand coalesced vs. non‑coalesced memory access and shared‑memory tiling.

Reviewer: ```suggestion
  - ``samples/2_Cookbook/0_MatrixTranspose``: Demonstrates implementing and optimizing a matrix transpose kernel. Helps to understand coalesced versus non‑coalesced memory access and shared‑memory tiling.
```

  - ``samples/2_Cookbook/17_llvm_ir_to_executable``: Shows how to take LLVM IR and turn it into a GPU executable.
  - ``samples/2_Cookbook/18_cmake_hip_device``: Demonstrates device‑side compilation with HIP in CMake. Useful for multi‑file device code.
  - ``samples/2_Cookbook/19_cmake_lang``: Shows how to use HIP as a first‑class CMake language.
  - ``samples/2_Cookbook/21_cmake_hip_cxx_lang``: Demonstrates mixing HIP with C++ and Clang toolchains for hybrid CPU/GPU projects.
  - ``samples/2_Cookbook/22_cmake_hip_lang``: Shows how to use the hip-lang CMake integration for advanced CMake setups and cross‑platform HIP builds.
  - ``samples/2_Cookbook/23_cmake_hiprtc``: Demonstrates HIP runtime compilation (HIPRTC) using CMake.

  .. note::

    There are no ``2_XXX`` or ``20_XXX`` code samples in the Cookbook. ``22_cmake_hip_lang`` is not built as part of the batch build process, and must be manually built. 

Building samples
----------------

To build all the hip-tests/samples use the following commands:

.. code:: bash

  cd hip-tests
  mkdir -p build && cd build
  cmake ../samples
  make build_samples

> Alternatively, to build specific samples use the following commands rather than ``build_samples``:

Reviewer: ```suggestion
Alternatively, to build specific samples, use the following commands rather than ``build_samples``:
```
